aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -196,16 +201,19 @@ Tensor empty_cpu(IntArrayRef size, const TensorOptions& options_, c10::optional<
       allocator,
       /*resizeable=*/true);
 
+  c10::DispatchKey dispatch_key = cpu ? at::DispatchKey::CPU : at::DispatchKey::CUDA;
   auto tensor = detail::make_tensor<TensorImpl>(
-      std::move(storage_impl), at::DispatchKey::CPU, dtype);
+      std::move(storage_impl), dispatch_key, dtype);
   // Default TensorImpl has size [0]
   if (size.size() != 1 || size[0] != 0) {
     tensor.unsafeGetTensorImpl()->set_sizes_contiguous(size);
   }
 
   auto memory_format = options.memory_format_opt().value_or(MemoryFormat::Contiguous);
-  tensor.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
-
+  //for contiguous format restride was done in set_sizes_contiguous, restride cases that did not go there
+  if ((size.size()==1 && size[0]==0) || memory_format != MemoryFormat::Contiguous) {
+      tensor.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
+  }
```

Comment:
Oof, this is so ugly

Reply:
Agreed, I'll see if I can get rid of `size` part of this conditional.